cooked-validators/src/Cooked/MockChain/Search.hs

```diff
@@ -0,0 +1,73 @@
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+
+module Cooked.MockChain.Search where
+
+import Control.Monad
+import Cooked.MockChain.BlockChain
+import Cooked.Output
+import qualified Ledger.Tx as Ledger
+import ListT (ListT (..))
+import qualified ListT
+import Optics.Core
+import qualified Plutus.V2.Ledger.Api as Pl2
+
+-- * The type of UTxO searches
+
+-- | If a UTxO is a 'TxOutRef' with some additional information, this type
+-- captures a "stream" of UTxOs.
+type UtxoSearch m a = ListT m (Pl2.TxOutRef, a)
+
+-- | Given a UTxO search, we can run it to obtain a list of UTxOs.
+runUtxoSearch :: Monad m => UtxoSearch m a -> m [(Pl2.TxOutRef, a)]
+runUtxoSearch = ListT.toList
+
+-- | Search all currently known 'TxOutRef's together with their corresponding
+-- 'TxInfo'-'TxOut'.
+allUtxosSearch :: MonadBlockChain m => UtxoSearch m Pl2.TxOut
+allUtxosSearch = allUtxos >>= ListT.fromFoldable
+
+-- | Like 'allUtxosSearch', but returns a Ledger-level representation of the
+-- transaction outputs, which might contain more information.
+allUtxosLedgerSearch :: MonadBlockChain m => UtxoSearch m Ledger.TxOut
+allUtxosLedgerSearch = allUtxosLedger >>= ListT.fromFoldable
+
+-- | Search all 'TxOutRef's at a certain address, together with their
+-- 'TxInfo'-'TxOut'.
+utxosAtSearch :: MonadBlockChainBalancing m => Pl2.Address -> UtxoSearch m Pl2.TxOut
+utxosAtSearch = utxosAt >=> ListT.fromFoldable
+
+-- | Like 'utxosAtSearch', but returns a Ledger-level representation of the
+-- transaction outputs, which might contain more information.
+utxosAtLedgerSearch :: MonadBlockChainBalancing m => Pl2.Address -> UtxoSearch m Ledger.TxOut
+utxosAtLedgerSearch = utxosAtLedger >=> ListT.fromFoldable
```

Comment:
Since the pattern seems to be:
* retrieve all utxos
* filter them until satisfied

I would think these function should be filters instead of "search" primitives.
Maybe it's not possible as `allUtxos` is only present in our `BlockChain` though, to be discussed